app/models/spree/adjustment.rb

```diff
@@ -35,10 +35,12 @@ class Adjustment < ActiveRecord::Base
     # So we don't need the option `dependent: :destroy` as long as
     # AdjustmentMetadata has no destroy logic itself.
     has_one :metadata, class_name: 'AdjustmentMetadata'
+    has_many :adjustments, as: :adjustable, dependent: :destroy
```

Comment:
I'm not following. Why would an adjustment have adjustments :thinking: ?

Reply:
Okay, this bit needs some explanation.

So this issue is: in Spree, adjustments are never taxable. In OFN, enterprise fee adjustments are taxable objects.

Previously when recording tax on an enterprise fee we were calculating it when the fee was created but not creating a tax adjustment. Instead we would just stick a tax amount in the `included_tax` column on the enterprise fee. This meant there was no proper link to a tax rate object and the tax couldn't be recalculated or updated. The only way to update it was to completely destroy all enterprise fees for an order and then re-create them all from scratch, via the `#recreate_all_fees!` method. It also avoids the correct (and slightly complex) logic in the `TaxRate` class and spreads the logic (and responsibility) for taxing things around the application.

Another issue was that with additional taxes (US and CA), we wouldn't even record that amount in the `included_tax` at all, and tax totals for each rate on the order were added to the order object (but with no link to which items had been taxed). So if there was a tax amount for say $5, you couldn't tell if it applied to line items, fees, shipments, or a combination.

These jumbled different methods of recording tax are OFN-specific and were added around 2014-2015 as temporary shortcuts to getting some of the tax features of Spree 2 (like having individual taxes recorded on line items), back when OFN was still on Spree 1. Those temporary shortcuts have had consequences though, and made a bit of a mess.

One of the reasons we can't go for "Option 2" in this change here is that the data basically isn't sufficient to recreate the fees and taxes. Until recently EnterpriseFee objects (the fee itself, not the adjustments it creates) were not soft-deleted. This means that lots of enterprise fee Adjustments in the database are now associated to EnterpriseFee objects that no longer exist. There's not enough information to reason about which tax rates were used when they were created or what amounts of tax are applied.

Reply:
So... when we create proper tax adjustments, they have a few requirements. Firstly the object being taxed must respond to `#tax_category`, which lets us figure out what rate(s) to apply. Secondly, the taxable object must respond to `#amount`, which is used in the calculation.

I've been trying to look at ways to do this, and this seems to be the simplest. So: yes, adjustments can have adjustments. In this case we could have an enterprise fee adjustment with an amount of $5 and a certain tax category. To tax it, we pass it to `Spree::TaxRate#adjust`. It'll add a tax adjustment on it which records the amount of tax, the rate used, whether it's additional or included etc, and we then have a tax adjustment associated to the fee adjustment. That association is already polymorphic (the "adjustable"). So in this case the "adjustable" object for tax recorded on a fee is: the fee (an adjustment).

Does that make sense.?

Reply:
I know it sounds a bit wild, but it works quite well with very minimal changes. I couldn't really see alternative options for the problem that we need to tax objects that are (fee) _adjustments_, and in order to tax things correctly we need to _add an associated tax adjustment_ to that object.

Reply:
This is the main bit of adapting needed to bringing the tax and adjustments changes from Spree 2.2. Everything else fits in quite well, but this needed an OFN-specific solution. It would be good to get @mkllnk 's thoughts on this as well.

Reply:
Another consequence of _not_ recording amounts of tax in proper tax adjustments (which means we have tax amounts attached to objects, but without a reference to a tax rate) is the madness that goes on with the `TaxRateFinder` class, and if you read the code comments [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/a1317be19b76d1953e38f9e47cbc8e5304d275e6/app/services/tax_rate_finder.rb#L59-L75) you'll see why I want to delete that class as soon as possible. This change brings us closer to be able to do that :fire:

Reply:
> I couldn't really see alternative options for the problem that we need to tax objects that are (fee) adjustments

The only other idea I would have is to make enterprise fees their own object which then has adjustments. But I can't think of any benefit of this except being more explicit and having less polymorphic complexity.

Your explanation of adjustable adjustments sounds reasonable and like a good way forward. I, too, want to get rid of the TaxRateFinder which should have been named InsufficientTaxRateGuesser.

Reply:
Well, it finds tax rates, just not always the right ones.

Reply:
> The only other idea I would have is to make enterprise fees their own object which then has adjustments.

Hmm... yeah that's not a bad idea. They would have to behave almost exactly like adjustments... I think if we went down that road, the data migration we would need in order to convert all existing fee adjustments to a different kind of object (with a different database table?) might end up being totally crazy. :grimacing:

Reply:
yes, it sounds like the other reasonable thing to do, but we could also wait and see if there's any extra value in going down that path. It seems that it would just add duplication now. The complexity of the polymorphism would still be there.